### ISSUE TITLE
#define ⇒ enum?

### ISSUE THREAD

**Issue by elfring** [bitdefender/bddisasm#6]
Would you like to replace more defines for constant values by [enumerations](https://en.wikipedia.org/wiki/Enumerated_type#C_and_syntactically_similar_languages "Description for enumerated types") to stress their relationships?
- [`ND_PREF_CODE_`…](https://github.com/bitdefender/bddisasm/blob/698ba367a105d9c0db637f29a6b7dedc92ca4ec4/bddisasm/include/prefixes.h#L8)
- [`ND_VEND_`…](https://github.com/bitdefender/bddisasm/blob/752bc626c4cfc1541441c5e507d65c3cde54f3ff/inc/bddisasm.h#L21)

**Comment by vlutas** [comment]
Sounds good, we will look into it. Thanks!

**Comment by vlutas** [comment]
Thinking about this, we could create **enums** for the indicated fields, but the reasoning behind our initial decision is the **INSTRUX** dimension - converting these fields to enums would mean that instead of 4 bits per field, we would use much more storage space, thus increasing the overall INSTRUX size and the memory footprint. 

Currently, the 11 fields x 4-bit/field are occupying less then a full QWORD, whereas switching to enums would increase this size to 11 x 32 (assuming 32 bit wide enums), which is 352 bits (44 bytes). 

In conclusion, we will let these fields as bit-fields for now, with #defines associated.

**Comment by elfring** [comment]
Can such enumerations be applied without specifying them as data structure members?

**Comment by vlutas** [comment]
Yes, they can - for example, you could keep the bitfield declaration inside the **INSTRUX** structure, and using **enum** instead of having **#define ND_CODE_...**. However, this does not improve readability in a significant way, since the **enum** fields are tied to bitfields inside the **INSTRUX** (not to a field of that enum's type), and the relationship between them is still not obvious. This, combined with a small number of **#defines** (and the reasons explained previously) makes us want to keep the current format.

**Comment by cbsirb** [comment]
But you can have a enum type as a bitfield: https://godbolt.org/z/v11zh4 (you even get a warning if the number of bits isn't enough to keep the max value in the enum).

**Comment by vlutas** [comment]
I know that, what I'm trying to say is that there's no point in defining the enum, if the type per se is not used in **INSTRUX**.
